File: srategy-design-pattern/code.py
from abc import ABC, abstractmethod


class PaymentStrategy(ABC):
    @abstractmethod
    def pay(self):
        pass


class PaypalPayment(PaymentStrategy):
    def pay(self, amount):
        print("Paid amount via paypal", amount)


class StripePayment(PaymentStrategy):
    def pay(self, amount):
        print("Paid amount via stripe", amount)


class PaymentProcessor:
    def __init__(self, paymentStrategy: PaymentStrategy):
        self.paymentStrategy = paymentStrategy
        
    def processPayment(self, amount):
        self.paymentStrategy.pay(amount)
        
paypal = PaymentProcessor(PaypalPayment())
stripe = PaymentProcessor(StripePayment())

paypal.processPayment(100)
stripe.processPayment(100)

File: strategy-design-pattern/code.js
class PaymentStrategy {
  pay(amount) {
    throw new Error("Method 'pay()' must be implemented.");
  }
}

class PayPalPayment extends PaymentStrategy {
  pay(amount) {
    console.log(`Paid ${amount} via PayPal`);
  }
}

class StripePayment extends PaymentStrategy {
  pay(amount) {
    console.log(`Paid ${amount} via Stripe`);
  }
}

class BitcoinPayment extends PaymentStrategy {
  pay(amount) {
    console.log(`Paid ${amount} via Bitcoin`);
  }
}

class PaymentProcessor {
  constructor(strategy) {
    this.strategy = strategy;
  }

  processPayment(amount) {
    this.strategy.pay(amount);
  }
}

const paypalProcessor = new PaymentProcessor(new PayPalPayment());
const stripeProcessor = new PaymentProcessor(new StripePayment());
const bitcoinProcessor = new PaymentProcessor(new BitcoinPayment());

paypalProcessor.processPayment(100);
stripeProcessor.processPayment(200);
bitcoinProcessor.processPayment(300);
